### ISSUE TITLE
Incorrect gradient for masked_select when inputs are non-contiguous

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#99638]
### 🐛 Describe the bug

Code:

```python
import torch

batch_dim = 3
in_dim = 5
classes_dim = 5
time_dim_sizes = torch.tensor([4, 3, 2], dtype=torch.int32)
time_dim = 4


def pack_padded_masked_select(x: torch.Tensor, sizes: torch.Tensor) -> torch.Tensor:
    """
    :param x: [B,T,...]
    :param sizes: [B]
    :return: [sum(sizes),...]
    """
    batch_dim, time_dim, *remaining_dims = x.shape
    mask = torch.arange(time_dim, device=x.device)[None, :] < sizes[:, None]  # [B,T]
    mask: torch.Tensor
    mask_bc = mask.view(batch_dim, time_dim, *[1] * len(remaining_dims))  # [B,T,...]
    # This, together with convolution, will cause a bad gradient?
    packed = torch.masked_select(x, mask_bc)
    packed = packed.view(-1, *remaining_dims)
    return packed


def pack_padded_index_select(x: torch.Tensor, sizes: torch.Tensor) -> torch.Tensor:
    """
    :param x: [B,T,...]
    :param sizes: [B]
    :return: [sum(sizes),...]
    """
    batch_dim, time_dim, *remaining_dims = x.shape
    mask = torch.arange(time_dim, device=x.device)[None, :] < sizes[:, None]  # [B,T]
    mask: torch.Tensor
    mask = mask.reshape(-1)  # [B*T]
    packed = torch.index_select(x.reshape(-1, *remaining_dims), dim=0, index=mask.reshape(-1).nonzero().flatten())
    return packed


def loss_packed_masked_select(logits: torch.Tensor, targets: torch.Tensor, sizes: torch.Tensor):
    logits_packed = pack_padded_masked_select(logits, sizes)
    targets_packed = pack_padded_masked_select(targets, sizes)
    loss_pt_packed_raw = torch.nn.CrossEntropyLoss(reduction="sum")(
        logits_packed, targets_packed.long()
    )
    return loss_pt_packed_raw


def loss_packed_index_select(logits: torch.Tensor, targets: torch.Tensor, sizes: torch.Tensor):
    logits_packed = pack_padded_index_select(logits, sizes)
    targets_packed = pack_padded_index_select(targets, sizes)
    loss_pt_packed_raw = torch.nn.CrossEntropyLoss(reduction="sum")(
        logits_packed, targets_packed.long()
    )
    return loss_pt_packed_raw


def loss_packed(logits: torch.Tensor, targets: torch.Tensor, sizes: torch.Tensor):
    logits_packed = torch.nn.utils.rnn.pack_padded_sequence(
        logits, sizes, batch_first=True, enforce_sorted=False
    )
    targets_packed = torch.nn.utils.rnn.pack_padded_sequence(
        targets, sizes, batch_first=True, enforce_sorted=False
    )
    loss = torch.nn.CrossEntropyLoss(reduction="sum")(
        logits_packed.data, targets_packed.data.long()
    )
    return loss


def loss_padded(logits: torch.Tensor, targets: torch.Tensor, sizes: torch.Tensor):
    loss = torch.nn.CrossEntropyLoss(reduction="none")(logits.transpose(1, 2), targets.long())  # [B,T]
    mask = torch.arange(time_dim, device=loss.device)[None, :] < sizes[:, None]  # [B,T]
    loss = torch.where(mask, loss, torch.zeros_like(loss))
    loss = loss.sum()
    return loss


try:
    import better_exchook
    better_exchook.install()
except ImportError:
    pass


try:
    import lovely_tensors as lt
    lt.monkey_patch()
except ImportError:
    pass


for loss_fn in [loss_packed, loss_padded, loss_packed_index_select, loss_packed_masked_select]:
    print("***", loss_fn.__name__)
    torch.manual_seed(42)

    net = torch.nn.Conv1d(in_dim, classes_dim, 5, padding="same")

    inputs = torch.randn(batch_dim, time_dim, in_dim, requires_grad=True)
    targets = torch.randint(0, classes_dim, (batch_dim, time_dim))

    x = inputs.transpose(1, 2)
    x = net(x)
    x = x.transpose(1, 2)

    loss = loss_fn(x, targets, time_dim_sizes)
    print("loss:", loss)
    print("bias grad:", torch.autograd.grad(loss, net.bias, create_graph=True))
```

Output:
```
*** loss_packed
loss: tensor grad NllLossBackward0 15.721
bias grad: (tensor[5] x∈[-2.044, 1.539] μ=0. σ=1.628 grad ConvolutionBackwardBackward0 [-1.437, 0.733, 1.539, 1.208, -2.044],)
*** loss_padded
loss: tensor grad SumBackward0 15.721
bias grad: (tensor[5] x∈[-2.044, 1.539] μ=4.768e-08 σ=1.628 grad ConvolutionBackwardBackward0 [-1.437, 0.733, 1.539, 1.208, -2.044],)
*** loss_packed_index_select
loss: tensor grad NllLossBackward0 15.721
bias grad: (tensor[5] x∈[-2.044, 1.539] μ=0. σ=1.628 grad ConvolutionBackwardBackward0 [-1.437, 0.733, 1.539, 1.208, -2.044],)
*** loss_packed_masked_select
loss: tensor grad NllLossBackward0 15.721
bias grad: (tensor[5] x∈[-1.277, 0.970] μ=3.576e-08 σ=0.889 grad ConvolutionBackwardBackward0 [0.712, -0.220, -1.277, 0.970, -0.184],)
```

As you see, all loss functions should behave the same. They also do return the same loss value. However, the gradient seems to be wrong for the case with `masked_select`.

Strangely, when I look at the gradient w.r.t. the logits, it seems still the same?

Strangely, when using `Linear` instead of `Conv1d`, I don't see the problem?

Strangely, if I add this code to the top of `pack_padded_masked_select`, so right before the `masked_select`, the gradient is correct?
```python
x = x.clone(memory_format=torch.contiguous_format)  # Fixes the gradient?
```

### Versions

PyTorch version: 1.13.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.6.3 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:24:02)  [Clang 11.1.0 ] (64-bit runtime)
Python platform: macOS-12.6.3-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] evotorch==0.1.0
[pip3] lovely-numpy==0.2.8
[pip3] mypy==0.991
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.2
[pip3] torch==2.0.0.dev20230209
[pip3] torchaudio==2.0.0.dev20230208
[pip3] torchdata==0.6.0.dev20230209
[pip3] torchvision==0.15.0.dev20230209
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mikaylagawarecki

**Comment by soulitzer** [comment]
Thanks for the report. The gradient is indeed wrong, but the issue is not with convolution, the incorrect gradients are computed when input to `loss_packed_masked_select` is a non-contiguous tensor. That's why calling `x = x.clone(memory_format=torch.contiguous_format) ` on it fixes it.

I have a smaller repro:
```python
import torch

batch_dim = 3
in_dim = 5
classes_dim = 5
time_dim_sizes = torch.tensor([4, 3, 2], dtype=torch.int64)
time_dim = 4

def pack_padded_masked_select(x: torch.Tensor, sizes: torch.Tensor) -> torch.Tensor:
    """
    :param x: [B,T,...]
    :param sizes: [B]
    :return: [sum(sizes),...]
    """
    batch_dim, time_dim, *remaining_dims = x.shape
    mask = torch.arange(time_dim, device=x.device)[None, :] < sizes[:, None]  # [B,T]
    mask: torch.Tensor
    mask_bc = mask.view(batch_dim, time_dim, *[1] * len(remaining_dims))  # [B,T,...]
    # This, together with convolution, will cause a bad gradient?
    packed = torch.masked_select(x, mask_bc)
    packed = packed.view(-1, *remaining_dims)
    return packed

def loss_packed_masked_select(logits: torch.Tensor, targets: torch.Tensor, sizes: torch.Tensor):
    logits_packed = pack_padded_masked_select(logits, sizes)
    targets_packed = pack_padded_masked_select(targets, sizes)
    loss_pt_packed_raw = torch.nn.CrossEntropyLoss(reduction="sum")(
        logits_packed, targets_packed.long()
    )
    return loss_pt_packed_raw

inputs = torch.randn(batch_dim, time_dim, in_dim, requires_grad=True, dtype=torch.float64)
targets = torch.randint(0, classes_dim, (batch_dim, time_dim))

def fn(input):
    x = input.transpose(1, 2).contiguous().transpose(1, 2)
    return loss_packed_masked_select(x, targets, time_dim_sizes)

torch.autograd.gradcheck(fn, (inputs,))
```

**Comment by ngimel** [comment]
masked_scatter_ implementation is wrong for non-contiguous self
